ai-in-directus-flows/index.md

```diff
@@ -0,0 +1,154 @@
+
+The emails sent to players are customised by directly editing the [liquid][8] template [files][9] provided by Directus. Only upon successfully completing the invite procedure does a user get provided with an active Directus user account and can hence access the game.
+
+## Tracking OpenAI Tokens
+AIVenture uses the [directus-extension-ai-pack][10] extension to interact with Open AI’s **Chat Completion API** within Directus Flows. The extension requires minimal setup before allowing chat prompts to be sent to Open AI and have responses returned for further processing in our flows.
+
+Chat GPT charges the API account owner for usage via [tokens][11] allocated by the length of your prompts and the resulting responses. In order to keep Open AI bills under control AIVenture allocates daily, per user, token limits on a player’s progress through the game as well as a hard limit that indicates the game needs to reach a conclusion.
+
+The Chat GPT API returns a token count with each response, however this was originally not passed on by [directus-extension-ai-pack][12]. Luckily, the extension is open-source and [a small PR later][13] we were able to access the Chat GPT usage data in our flow:
+
+```json
+"usage": {
+    "completion_tokens": 17,
+    "prompt_tokens": 57,
+    "total_tokens": 74
+} 
+```
+
+Each prompt/response combination is stored in a Directus collection and various query filters and aggregators are used to provide the game engine with the total number of tokens used by a player in a day and the total tokens used for the current game.
+
+If the player reaches their daily limit, no more Chat GPT prompts can be sent to the API and the game will prompt you to wait until tomorrow before being able to progress.
+
+Once the total game token limit is reached the prompt sent to Chat GPT is modified such that the next responses will include a satisfying conclusion to the game currently in progress.
+
+## The game engine
+The [AIVenture][14] game engine is responsible for:
+
+1. Providing initial context to Chat GPT (eg. “You are a text adventure game writer…”).
+2. Generating an introduction (eg. “You are standing in an open field…”).
+3. Responding to player prompts in a way compliant with any game rules and the initial context.
+4. Managing Chat GPT interactions including token counts and unexpected responses.
+5. Providing a satisfactory end game state.
+6. State management for games in progress.
+
+In AIVenture all this logic is represented in Directus Flows. However with the large number of operations required we need to plan our flow architecture carefully to avoid unnecessarily complicating our game engine flow.
```

Comment:
```suggestion
In AIVenture, all of this logic is represented in Directus Flows. However with the large number of operations required we need to plan our flow architecture carefully to avoid unnecessarily complicating our game engine flow.
```